Fix genesis init for migrated chains

piersy opened the pull request:
When the celo blockchain launched it did so without `gasLimit`, `difficulty`
and `uncleHash` fields on the celo block.

In the gingerbread hardfork these fields were added to the block, to improve
celo's compatibility with ethereum, and a zero (or not) `uncleHash` was used to
distinguish the old and new block formats.

When transitioning to the op-stack we discovered that these fields are set to
default values during genesis init if they are unset. This broke our approach
to encoding, so to prevent this happening we would only set defaults in a
migrated chain context (where cel2 time is set an non zero), this allowed us to
skip setting defaults for migrated chains but retain the defaults for tests
which rely heavily on them.

In the op-stack we were using a zero gas limit as an indication of what type of
block encoding to use, but this presented it's own problems for some tests that
used blocks without a gas limit set, so we subsequently switched (#300) to
having an explicit flag set inside headers to specify which encoding to use.
When loading blocks from leveldb the RLP structure could be introspected to
determine the type of block that was being loaded, our thinking was that for
all migrated blocks we could rely on this approach to setting the encoding, and
for any new blocks they would all be being created after the cel2 fork and so
would use the most recent encoding. We removed the special casing of default
setting in `toBlockWithRoot` and relied on the explicit flag.

Unfortunately we forgot about what happens at genesis init, where `NewBlock` is
called to create the genesis block. With the special casing removed, the
defaults being set and `NewBlock` always producing blocks that use the new
encoding we discovered that initing the genesis on a migrated chain resulted in
a different genesis block.

To fix this the special casing has been re-introduced in `toBlockWithRoot` and
extended to prevent default setting of `uncleHash` inside `NewBlock` when
operating in the context of a migrated chain. Also if the genesis block is a
pre-gingerbread block it is marked as such so that it is encoded correctly.

There's was one remaining issue which is that `TestFeeHistory`, initialises a
pre-gingerbread genesis but expects it to have the new block format, so to
account for that an extra condition was added to only use the old format for
pre-gingerbread blocks that are being initialised as a result of executing the
initGenesis command.

# Tested
Locally and got the same 0 block as returned by https://baklava-forno.baklava.celo-testnet.org/ also saw a snap sync node get past syncing beacon headers stage.

Also added unit tests to verify the genesis block created under different conditions.

Resolves https://github.com/celo-org/celo-blockchain-planning/issues/900

piersy committed:
Handle migrated chains during genesis init

We do not set defaults for migrated chains and we also ensure that if
the genesis block is pre-gingerbread it is marked as such.

piersy committed:
Add initing genesis condition

palango commented on core/celo_genesis.go:
```diff
@@ -167,3 +167,13 @@ func addFeeCurrencyToStorage(feeCurrencyAddr common.Address, oracleAddr common.A
 	storage[incHash(structStart, 1)] = common.BigToHash(big.NewInt(int64(FeeCurrencyIntrinsicGas))) // intrinsicGas
 }
+
+// SetInitingGenesis sets marks this genesis as one that is being used in the initGenesis operation.
```
```suggestion
// SetInitingGenesis marks this genesis as one that is being used in the initGenesis operation.
```

palango commented on core/genesis.go:
```diff
@@ -589,33 +593,50 @@ func (g *Genesis) ToBlock() *types.Block {
-		head.Difficulty = params.GenesisDifficulty
+	number := new(big.Int).SetUint64(g.Number)
+	head := &types.Header{}
+	// If we are in an initGenesis flow and this is a pre gingerbread block,then
```
```suggestion
	// If we are in an initGenesis flow and this is a pre-gingerbread block, then
```

palango commented on core/genesis.go:
```diff
@@ -589,33 +593,50 @@ func (g *Genesis) ToBlock() *types.Block {
+	head := &types.Header{}
+	// If we are in an initGenesis flow and this is a pre gingerbread block,then
+	// set the block to be pre-gingerbread. This affects the way it is encoded,
+	// which affects it's hash. Ideally we could do without the InitingGenesis
```
```suggestion
	// which affects its hash. Ideally we could do without the InitingGenesis
```

palango reviewed (commented):
Looks good, nice PR description. Comments are typo nits. 

I guess that fixes the snap sync problems, does it also solve the `getBlock(0)` problems that Alec mentioned?

piersy committed:
Update core/celo_genesis.go

Co-authored-by: Paul Lange <palango@users.noreply.github.com>

piersy committed:
Update core/genesis.go

Co-authored-by: Paul Lange <palango@users.noreply.github.com>

piersy committed:
Update core/genesis.go

Co-authored-by: Paul Lange <palango@users.noreply.github.com>

piersy commented:
> Looks good, nice PR description. Comments are typo nits.
> 
> I guess that fixes the snap sync problems, does it also solve the `getBlock(0)` problems that Alec mentioned?

Sorry meant to mention that I tested this locally and got the same 0 block as returned by https://baklava-forno.baklava.celo-testnet.org.

Given that they share the same block there shouldn't be any problem snap syncing. I currently have a node that is snap syncing but its still at the beacon headers sync stage.

alecps commented on core/genesis.go:
```diff
@@ -589,33 +593,50 @@ func (g *Genesis) ToBlock() *types.Block {
+
+	// When initialising the genesis we don't want to override unset gasLimit or
+	// difficulty fields since these fields were not present at the celo chain genesis.
+	if !g.Config.IsMigratedChain() {
```
Just double checking, should this if-statement also include g.InitingGenesis?

piersy replied:
There are lots of tests that we have added that make use of migrated chains, adding the `InitingGenesis` constraint breaks them.

piersy replied:
I've now replaced  `InitGenesis` with a call to `testing.Testing`

alecps commented on core/genesis.go:
```diff
@@ -589,33 +593,50 @@ func (g *Genesis) ToBlock() *types.Block {
 		if g.BaseFee != nil {
 			head.BaseFee = g.BaseFee
 		} else {
+			// If defaults are not ignored, set default values for base fee.
```
Just to clarify my own understanding, how do we know defaults are not ignored at this point in the code?

piersy replied:
This is a stray comment from a previous version, will remove 👍

alecps commented on core/types/block.go:
```diff
@@ -286,7 +287,7 @@ func NewBlock(header *Header, body *Body, receipts []*Receipt, hasher TrieHasher
 	}
 
-	if len(uncles) == 0 {
+	if len(uncles) == 0 && !bType.IsMigratedChain() {
```
Do we maybe want the EmptyUncleHash for new blocks in a migrated chain that aren't the genesis?

piersy replied:
Probably we do 👍

piersy replied:
Well spotted, looks like that could have broken future block generation, but strange that it didn't cause any tests to fail.

alecps reviewed (approved):
Looks great! Just left a few questions

piersy committed:
Pr fixes

piersy committed:
Remove the use of initingGenesis flag

This flag was being used to indicate the context that we were executing
in, in this case a  non test production context. Instead we can use
testing.Testing() that tells us whether we are operating in a test
context.

piersy committed:
Re-word comment

alecps commented on core/types/block.go:
```diff
@@ -286,7 +288,9 @@ func NewBlock(header *Header, body *Body, receipts []*Receipt, hasher TrieHasher
-	if len(uncles) == 0 {
+	// We prevent setting the EmptyUncleHash only when we are on a migrated
+	// chain and the block is pre-gingerbread.
+	if len(uncles) == 0 && !(bType.IsMigratedChain() && !bType.IsGingerbread(header.Number)) {
```
I think this new logic prevents setting the EmptyUncleHash only when on a migrated chain and the block is post gingerbread

alecps replied:
Oh sorry, think I missed the parenthesis

piersy committed:
Revert "Remove the use of initingGenesis flag"

This reverts commit df753d9f2b18a7954da4ccdd049923ed9e4f31db.

piersy committed:
Add tests of toBlockWithRoot

Adds tests to check that toBlockWithRoot correctly does or doesn't set
defaults depending on the genesis config.

Note the condition around setting the unclehash in NewBlock was changed
as a result of this, since before even if it avoided setting the empty
uncle hash the else case would calculate the hash for zero uncles, still
resulting in the empty uncles hash.

alecps commented on core/types/block.go:
```diff
@@ -288,15 +288,17 @@ func NewBlock(header *Header, body *Body, receipts []*Receipt, hasher TrieHasher
+		if len(uncles) == 0 {
+			b.header.UncleHash = EmptyUncleHash
+		} else {
+			b.header.UncleHash = CalcUncleHash(uncles)
```
I guess uncles must be empty if we **are** pre-gingerbread and on a migrated chain, so we don't have to copy uncles or set unclehash for that case?

piersy replied:
Yep, that is the logic, is it clear from the comment?

alecps replied:
Yes it's clear, just wanted to double check

alecps reviewed (approved):
Looks great!
